Add mapping for Roland DJ-505

Holzhaus opened the pull request:
# Roland DJ-505

The Roland DJ-505 is an all-in-one USB MIDI controller with an integrated audio interface. It has controls for 2 decks that can be toggled between decks to play with 4 decks. It also features a TR-S step sequencer with sounds from the TR-808 and the TR-909 drum machines (TR-606 and TR-707 are available via firmware update). It's positioned between the compact [DJ-202](https://www.mixxx.org/wiki/doku.php/roland_dj-202) (#1595, [Forum Thread](https://mixxx.org/forums/viewtopic.php?f=7&t=11664&sid=a2e4fc02a44adb4c09dcf6a099dea577))  and the flagship DJ-808 controller.

## Links
- [Manufacturer's Product Page](http://my.roland.com/products/dj-505/)
- [Wiki Entry](https://www.mixxx.org/wiki/doku.php/roland_dj-505)
- [Forum Thread](https://mixxx.org/forums/viewtopic.php?f=7&t=12688)

## Remaining Issues

- [x] Implement selecting Effects ~~via FX SELECT buttons~~ (just use standard FX unit mapping)
- [x] ~~Fix LEDs for BACK/ADD PREPARE buttons~~ Does not work in Serato either
- [x] VINYL button
- [x] PHONO CH Input Select (might work, not tested yet)
- [x] ~~There might be some bug in the version of Mixxx that I'm using which leaves some Sampler Slots unaffected by `engine.setValue()` calls~~ Works on latest master
- [x] Fix TR-S Sampler S7/S8 MIDI mapping
- [x] Fix Deck 3/4 toggling
- [x] Custom Hotcue color support (depends on PR #2030)
- [x] VU Meter support
- [ ] TR-S Syncing currently works at the BPM level, but phase syncing is not implemented yet. As a workaround, the NUDGE button can be used to adjust the phase. This depends on MIDI clock I/O ([Launchpad Bug #682221](https://bugs.launchpad.net/mixxx/+bug/682221)).
- [ ] Some performance pad modes are missing (Slicer [ [Launchpad Bug #1828886](https://bugs.launchpad.net/mixxx/+bug/1828886) ], Slicer Loop, Saved Loop, Flip)
- [ ] Buttons for sorting the library by BPM/Song/Artist/Key are not mapped since no Mixxx controls seem to exist for it ([Launchpad Bug #1828882](https://bugs.launchpad.net/mixxx/+bug/1828882), PR #2118)
- [ ] Some buttons are not mapped yet (e.g. BACK)
- [ ] LEDs on BACK/ADD PREPARE do not work (this seems to be a hardware/firmware bug and does not work in Serato either)
- [ ] Controller does not send current cross fader value on Serato Sysex Message (this seems to be a hardware/firmware bug and does not work in Serato either)
- [ ] Support for colored hotcues has been added, but depends on PR #2030 which has not been merged yet.

## Controller Overview
![Roland DJ-505](http://cdn.roland.com/assets/images/products/gallery/dj-505_t_gal.jpg)

Holzhaus committed:
Roland DJ-505: Copy over Roland DJ-202 mapping and use it as base

This copies the mapping for the Roland DJ-202 by Silvio Ankermann from
Lykos153/mixxx@87d0bbf and replaces all occurrences of "202" with "505".

Holzhaus committed:
Roland DJ-505: Fix a bunch of linter warnings

Holzhaus committed:
Roland DJ-505: Add more fixes

Holzhaus committed:
Roland DJ-505: Add some fixes

Holzhaus committed:
Roland DJ-505: Fix slip mode button

Holzhaus committed:
Roland DJ-505: Fix some issues

Holzhaus committed:
Roland DJ-505: Improve Browse Knob

Holzhaus committed:
Roland DJ-505: Add "Roll" mode performance pads

Holzhaus committed:
Roland DJ-505: Add "Loop" mode performance pads

Holzhaus committed:
Merge branch 'roland-dj-505-mapping' of github.com:Holzhaus/mixxx into roland-dj-505-mapping

Holzhaus committed:
Roland DJ-505: Improve XML mapping and add descriptions

Holzhaus committed:
Roland DJ-505: Implement support for cross fader settings

Holzhaus committed:
Roland DJ-505: Add basic support for CH 1/2 input switches

Holzhaus committed:
Roland DJ-505: Add basic support for the TR-S section (BPM sync)

Holzhaus committed:
Roland DJ-505: Add TODO comments to script code

Holzhaus committed:
Roland DJ-505: Add preliminary performance pad LED support

Holzhaus committed:
Roland DJ-505: Add support for TR/Sampler in Serato mode

Holzhaus committed:
Roland DJ-505: Add FX support for TR/Sampler

Holzhaus committed:
Roland DJ-505: Add more TODO comments

Holzhaus committed:
Roland DJ-505: Fix performance pads sampler mode

Holzhaus committed:
Roland DJ-505: Add initial support for custom TR-S samples

Holzhaus committed:
Roland DJ-505: Add JSHnt configuration

Holzhaus committed:
Roland DJ-505: Refactor PadSection code to use ComponentContainers

Holzhaus committed:
Roland DJ-505: Add parameter support for ROLL mode

Holzhaus committed:
Roland DJ-505: Add Velocity Sampler Mode for Performance Pads

Holzhaus committed:
Roland DJ-505: Add Cue Loop mode for Performance Pads

Holzhaus committed:
Roland DJ-505: Add Pitch Play Mode for Performance Pads

Holzhaus committed:
Roland DJ-505: Fix KEYLOCK/TEMPO RANGE Button

Holzhaus committed:
Roland DJ-505: Fix DJ505.setChannelInput function

Holzhaus committed:
Roland DJ-505: Fix Sampler volume on Velocity Sampler mode

Holzhaus committed:
Roland DJ-505: Fix DJ505.shiftButton Sampler handling

Holzhaus committed:
Roland DJ-505: Fix DJ505.crossfader.setCurve indentation

Holzhaus committed:
Roland DJ-505: Use standard EffectUnit component

When the Roland is in Serato Mode, the LED issue (i.e. button LEDs
disable themselves after release) does not occur anymore.

Holzhaus committed:
Roland DJ-505: Fix some linter errors in DJ505.init() function

Holzhaus committed:
Roland DJ-505: Fix some linter errors in DJ505.autoShowDecks

Holzhaus committed:
DJ-505: Fix linter errors in DJ505.crossfader

Holzhaus committed:
Roland DJ-505: Fix linter errors in DJ505.setChannelInput

Holzhaus committed:
Roland DJ-505: Fix linter errors in DJ505.Deck

Holzhaus committed:
Roland DJ-505: Fix linter errors in DJ505.SyncButton

Holzhaus committed:
Roland DJ-505: Fix linter errors in SlipModeButton

Holzhaus committed:
Roland DJ-505: Fix linter errors in DJ505.PadSection

Holzhaus committed:
Roland DJ-505: Apply FX to Sampler and custom TR-S samples

Holzhaus committed:
Roland DJ-505: Make sampler controls affect all 16 samplers instead of 8

Holzhaus committed:
Roland DJ-505: Add BACK/SONG and ADD PREPARE/ARTIST buttons

Holzhaus committed:
Roland DJ-505: Fix RELOOP/EXIT button behaviour

Holzhaus committed:
Roland DJ-505: Make Sync button behaviour similar to Owner's Manual

Holzhaus committed:
Roland DJ-505: Set sendShifted to true in effect unit

uklotzde commented:
Welcome, Jan. Would you please sign our [Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ)?

Nice work with a comprehensive documentation. You can ignore the Codacy issues about unused function arguments.

If the basic functions are working and if none of our mapping and JS wizards has any objections, it is mainly your decision when to merge this initial version of the mapping. It doesn't matter if some hardware controls are not mapped, yet. You may add the TODO check list from this PR to the wiki page if you want to keep it.

Holzhaus commented:
Thanks, I signed it.

Holzhaus committed:
Roland DJ-505: Rework deck switching mechanic

Holzhaus committed:
Roland DJ-505: Fix TR-S Sampler S7/S8 MIDI mapping

Holzhaus committed:
Roland DJ-505: Map VINYL button ([SHIFT]+[SLIP]) to vinylcontrol_enabled

Holzhaus committed:
Roland DJ-505: Fix TAP button

Holzhaus committed:
Roland DJ-505: Fix CUE/PLAY button shift values

Holzhaus committed:
Roland DJ-505: Remove obsolete PadSection.resetLEDs()

Holzhaus committed:
Roland DJ-505: Fix TEMPO RANGE button

Holzhaus committed:
Roland DJ-505: Fix some minor code style nits

Holzhaus committed:
Roland DJ-505: Map SLOT SELECT button to quantize on/off

Holzhaus committed:
Roland DJ-505: Map Hotcue Mode PARAMETER buttons to beats_translate

Holzhaus committed:
Roland DJ-505: Remove obsolete LED workaround for AUTO LOOP button

Holzhaus committed:
Roland DJ-505: Improve controller documentation

Holzhaus committed:
Roland DJ-505: Apply FX to Aux channels if LINE/PHONO switch is used

Holzhaus committed:
Roland DJ-505: Make Aux channels configurable

Holzhaus committed:
Roland DJ-505: Remove unused setting

Holzhaus committed:
Roland DJ-505: Use passthrough mode for PC/LINE/PHONO input selection

Holzhaus committed:
Roland DJ-505: Improve PadSection code readability by using JS enums

Holzhaus committed:
Roland DJ-505: Use Auxiliary1 input for TR-S by default

Holzhaus committed:
Roland DJ-505: Fix TR-S group config

Holzhaus committed:
Roland DJ-505: Fix some warnings during controller init

Holzhaus committed:
Roland DJ-505: Document hardware/firmware quirks in source code

Holzhaus committed:
Roland DJ-505: Add VU Meter support

Holzhaus committed:
Merge branch 'roland-dj-505-mapping' of github.com:Holzhaus/mixxx into roland-dj-505-mapping

Holzhaus commented:
IMHO this is ready to merge now. It would be nice to have this tested by somebody else, but it should work good enough. I added the remaining issues to the Wiki entry.

uklotzde commented:
Very well done. Together with the wiki page this is ready right from the start! @Be-ing Any thoughts?

First of all I'm not a JS developer. Just some remarks that came to my mind:
- I see many non-strict comparion operators ==/!= that could be replaced by ===/!==? Wondering why jshint didn't mention that.
- jshint gives some other warnings, not all of them might apply:
```
/tmp/Roland_DJ-505-scripts.js: line 170, col 49, Functions declared within loops referencing an outer scoped variable may lead to confusing semantics. (DJ505)
/tmp/Roland_DJ-505-scripts.js: line 941, col 48, ['roll'] is better written in dot notation.
/tmp/Roland_DJ-505-scripts.js: line 942, col 48, ['sampler'] is better written in dot notation.
/tmp/Roland_DJ-505-scripts.js: line 951, col 26, ['hotcue'] is better written in dot notation.
/tmp/Roland_DJ-505-scripts.js: line 957, col 26, ['cueloop'] is better written in dot notation.
/tmp/Roland_DJ-505-scripts.js: line 966, col 26, ['roll'] is better written in dot notation.
/tmp/Roland_DJ-505-scripts.js: line 975, col 26, ['sampler'] is better written in dot notation.
/tmp/Roland_DJ-505-scripts.js: line 978, col 26, ['velocitysampler'] is better written in dot notation.
/tmp/Roland_DJ-505-scripts.js: line 981, col 26, ['loop'] is better written in dot notation.
/tmp/Roland_DJ-505-scripts.js: line 984, col 26, ['pitchplay'] is better written in dot notation.
/tmp/Roland_DJ-505-scripts.js: line 1288, col 5, 'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
/tmp/Roland_DJ-505-scripts.js: line 1289, col 5, 'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
/tmp/Roland_DJ-505-scripts.js: line 1290, col 5, 'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
/tmp/Roland_DJ-505-scripts.js: line 1310, col 22, Functions declared within loops referencing an outer scoped variable may lead to confusing semantics. (DJ505)
/tmp/Roland_DJ-505-scripts.js: line 1343, col 20, Functions declared within loops referencing an outer scoped variable may lead to confusing semantics. (DJ505)
```

uklotzde commented on res/controllers/Roland_DJ-505-scripts.js:
```diff
@@ -0,0 +1,1405 @@
+            // Strip search.
+            var oldPos = engine.getValue(this.currentDeck, "playposition");
+            // Scale to interval [0,1].
+            newValue = newValue / 0xff;
```
The delta newValue will be in the range [-64, 63]. Dividing by 255 will not scale it to the unit interval as the comment pretends and is confusing.

Using just DJ505.stripSearchScaling as the scaling constant is sufficient. 50/255 equals ~0.2. Scaling by the actual duration is still needed to provide a linear feedback between motion/rotation and time.

Holzhaus replied:
Fixed in 31b62638cc7c9491548cbe0b70d38593d9aa5e5a. I used 0.15 instead of 0.2 as stripSearchScaling constant because that provides a bit more precision when searching through the track.

Holzhaus replied:
Pinging @Lykos123 and @upCASE because that issue is also present in the mapping for the Roland DJ-202 (original PR #1595, [updated version gist](https://gist.github.com/upCASE/88fb258e85546092af18bc3ff924f752#file-roland_dj-202-scripts-js)).

upCASE replied:
Thanks! Will fix this in my version and see if I find the time to finally upload this and create a PR for the 202 mapping.

Holzhaus committed:
Roland DJ-505: Use dot notation to fix some linter warnings

Holzhaus committed:
Roland DJ-505: Use strict comparisons

Holzhaus committed:
Roland DJ-505: Do not define PitchPlayMode pads inside loop

Holzhaus committed:
Roland DJ-505: Do not define CueLoopMode pads inside loop

Holzhaus committed:
Roland DJ-505: Use JS enums instead of const values (ECMAScript >= v6)

Holzhaus committed:
Roland DJ-505: Create alias for complete TR-S FX enableOnChannelButton

This fixes some JSHint warning regarding declaring functions in loops.

Holzhaus committed:
Roland DJ-505: Improve strip search implementation

Holzhaus commented:
> First of all I'm not a JS developer.

Me neither ;-)
> * I see many non-strict comparion operators ==/!= that could be replaced by ===/!==? Wondering why jshint didn't mention that.

You're right, I fixed it in b818531. I had to put `// jshint eqeqeq:true` at the top of the file to make jshint complain about them.

> * jshint gives some other warnings, not all of them might apply

Ok, these are also fixed. For some reason I eslint (which I was mistakenly using instead of jslint) did not warn about them.

Holzhaus committed:
Roland DJ-505: Improve VU meter / peak indicator LEDs

Holzhaus committed:
Roland DJ-505: Add support for custom hot cue colors

This depends on PR #2030. If the PR is merged, custom hot cue colors
will immediately work with this mapping. In the meantime, the pads will
light up in the mode's color (i.e. white in hot cue mode, blue in cue
loop mode).

Holzhaus committed:
Roland DJ-505: Prevent pitchplay pad connection if mode is not activated

Holzhaus committed:
Roland DJ-505: Remove invalid CUE/PFL button shift options

Holzhaus committed:
Roland DJ-505: Use ComponentContainers for all performance pad modes

Holzhaus committed:
Roland DJ-505: Add support for hotcue colors to Pitch Play Mode

Holzhaus commented:
I discovered and fixed some minor issues and also added Hotcue color support. It depens on PR #2030 (which has not been merged yet).

Holzhaus committed:
Roland DJ-505: Remove unused param from PitchPlayMode.outputColor func

Holzhaus committed:
Roland DJ-505: Add forum URL to mapping XML file

Holzhaus commented:
IMHO this can be merged now. If #2118 is merged first, it could also map the currently unmapped "Sort by BPM/Key/Artist/Song" buttons. If not, I could just open a separate PR for this.

uklotzde commented:
Does this mapping in its current state work even without #2030 and #2118? Or is it mandatory to merge those PRs first?

Holzhaus commented:
> Does this mapping in its current state work even without #2030 and #2118?
Yes. Without PR #2030, performance pads will then light up in the mode's color (White in Hotcue Mode, Blue in CueLoop mode). Hotcue colors in Pitch Play Mode will work regardless of PR #2030.

I didn't add anything related to #2118 to this mapping yet.

Holzhaus commented:
I added support for library sort buttons to a different branch. If #2118 is merged, we can just cherry-pick commit Holzhaus/mixxx@ba3e01f5d6f703a3b7fedf93dae0513299997d7c and add it to this PR, but if something block #2118, this mapping can still be merged as-is.

uklotzde commented:
LGTM

I'm not able to test it, but I don't see any reason for holding this back. The earlier it becomes available in the regular master builds the more likely other owners of this device will use and test the mapping.

Holzhaus commented:
Thanks for merging!

Be-ing commented:
Thanks for contributing this and the thorough documentation @Holzhaus. And thank you @uklotzde for merging this while I was busy.

I read through the wiki page for this controller and made some minor edits and reorganization. One suggestion I have for the mapping is to add `reloop_andstop`. I suggest mapping shift + autoloop for this.

I like that you used shift + PFL for beatgrid adjustment. Mappings typically don't have anything mapped for shift + PFL and controllers don't typically have a dedicated button for beatgrid adjustment.
